### DISCUSSION TITLE
Why is this code laggy? Standard keyboard input vs modifiers.

### DISCUSSION THREAD

**Discussion by Muzz** [emilk/egui#3339] [Q&A]
Quick question. 

I'm making a little painting program, and calling a custom painting function, and inside of it i have these two functions to change  the angle and panning of the canvas. 

For some reason if I put the dragging in a check for if the keyboard is pressed, then i get some weird behavior, where the mouse will move, but won't update the moving of the canvas until I've moved the mouse some distance. If i put it under a modifier check, it works fine. 

Any insight would be greatly appreciated!
        
```
 fn custom_painting(&mut self, ui: &mut egui::Ui) {
        if let Some(image_buffer) = &mut self.image_buffer {
            let (rect, response) =
                ui.allocate_exact_size(egui::Vec2::splat(800.0), egui::Sense::drag());

            // This is  very responsive
            let shift_down = ui.input(|i| i.modifiers.shift);
            if shift_down {
                image_buffer.angle += response.drag_delta().x * 0.01;

            }

            // this is unresponsive and i don't know why?
            else if ui.input(|i| i.key_pressed(egui::Key::Space)) {
                println!("pressed");
                image_buffer.x_loc += response.drag_delta().x * 0.01 ;
                image_buffer.y_loc += response.drag_delta().y * -0.01  ;
            }

            // Clone locals so we can move them into the paint callback:
            let angle = image_buffer.angle;
            let x_loc = image_buffer.x_loc;
            let y_loc = image_buffer.y_loc;
            let zoom = image_buffer.zoom;
            
            let rotating_triangle = image_buffer.rotating_triangle.clone();

            let callback = egui::PaintCallback {
                rect,
                callback: std::sync::Arc::new(egui_glow::CallbackFn::new(move |_info, painter| {
                    rotating_triangle.lock().paint(painter.gl(), angle,x_loc,y_loc,zoom);
                })),
            };
            ui.painter().add(callback);

        }
    }```

**Comment by Mingun** [comment]
This is because `else if`. If you press shift, then the code that detect of pressing space does not run

**Reply by Muzz** [reply]
This is incorrect, the issue is NOT that the code doesn't run when you press shift, that is intended behavior, it's that there is a bizarre delay between updating x_loc and y_loc when space is held down. 

the weirdest thing is though that a print statement within that block fires instantly when space is pressed.

**Comment by YgorSouza** [comment]
If I understood correctly, you should be using [key_down](https://docs.rs/egui/0.22.0/egui/struct.InputState.html#method.key_down) instead of [key_pressed](https://docs.rs/egui/0.22.0/egui/struct.InputState.html#method.key_pressed).

**Reply by Muzz** [reply]
I'll give that a shot, thank you!!

**Comment by emilk** [comment]
Are you calling `egui_ctx.request_repaint()` to ensure proper animation?
